keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1424 @@
+- 1 is for the separation character
+- 2*(10^4+10^4) is the number of indexes explicitly listed in the field - two indexes per gap.
+
+The size of the `failedIndexes` field is limited by: `(10+1)*10^4=0.105Mi`, where:
+- 10 is the maximal number of digits in the indexes,
+- 1 is for the separation character
+- 10^4 is the maximal number of indexes present in the field.
+
+Thus, the size of both fields is capped at `0.572Mi` for the limits in (1.) and
+`0.525Mi` for the limits in (2.).
+
+For comparison, before the introduction of `.status.failedIndexes`, the max
+size of the `.status.completedIndexes` was limited by `(5+1)*10^5*2/3=0.382Mi` in
+the (1.) case, and `(10+1)*2*10^4=0.21Mi` in the (2.) case. This means an increase
+of `0.19Mi`.
+
+The values of the limits are aligned with the values for the soft limits proposed
+as a fix for the for regular indexed jobs
+(see [here](https://github.com/kubernetes/kubernetes/issues/118085#issuecomment-1564520559)).
+However, in case when `backoffLimitPerIndex` is used we propose these limits
+to be hard.
+
+We believe that the scalability limits should be enough for most of Job use-cases.
+For workloads requiring larger jobs users should be able to create multiple Jobs,
+orchestrated by the [JobSet](https://github.com/kubernetes-sigs/jobset).
+
+Still, we leave it for reassessment for Beta if the limits are strong (error)
+or soft (warning).
+
+### Expotential backoff delay issue
+
+Currently, a pod is recreated by the Job controller with expotential backoff
+delay (10s, 20s, 40s ...), counted from the last failure time.
+
+One complication is that the last failure time for failed pods may increase with
+time, as it fallbacks to `now` in some cases
+(see in [code](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/backoff_utils.go#L160-L182)).
+Thus, there is a risk that due to the presence
+of pods hitting the fallback the last failure time is continuously bumped,
+thus shifting the time to recreate the pod.
+
+This risk is present both when computing the expotential backoff delay globally
+(as for regular indexed Jobs), or per-index as proposed in in this KEP
+(see [Expotential backoff delay per index](#expotential-backoff-delay-per-index)).
+
+In order to mitigate this risk currently the time of last failure is recorded
+in-memory (globally for all pods within a Job). And a new failed pod may bump
+it only until it is added to the `uncountedTerminatedPods` structure.
+
+However, tracking the last failure time per index might be costly for memory
+consumption (see [Expotential backoff delay with in-memory tracking](#expotential-backoff-delay-with-in-memory-tracking)).
+
+Thus, in order to mitigate this risk we propose to compute the finish time for
+a pod as min value of the following (avoiding the ever-increasing fallback to `now`):
+1. max `finishAt` of all containers, if specified for all containers
+2. `LastTransitionTime` for the `Ready=False` condition
```

Comment:
Does kubelet change the condition when the pod is marked with a deletionTimestamp?

Reply:
Yes, and this is what typically happens. Setting deletionTimestamp initiates the termination process by kubelet. Then containers are stopped, and finally the phase is changed to terminal (along with setting the condition). I was also thinking about checking the Reason field to be `PodFailed` or `PodCompleted` (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/generate.go#L234-L247). Probably can be figured out during implementation.